src/merkle/cmr.rs

```diff
@@ -54,11 +55,39 @@ impl CommitMerkleRoot for Cmr {
             CommitNodeInner::Fail(_, _) => Cmr::tag_iv(b"Simplicity-Draft\x1fCommitment\x1ffail"),
             CommitNodeInner::Hidden(h) => *h,
             CommitNodeInner::Jet(j) => Cmr::tag_iv(b"Simplicity-Draft\x1fJet").update_1(j.cmr()),
+            CommitNodeInner::Word(_) => Cmr::tag_iv(b"Simplicity-Draft\x1fIdentity"),
         }
     }
 }
 
 impl Cmr {
+    /// Compute the CMR of a word
+    ///
+    /// This is equal to the IMR of the equivalent scribe, converted to a CMR in
+    /// the usual way for jets.
+    // FIXME avoid calling `powers_of_two_vec`; also replace this "construct the scribe then
+    //  compute its CMR" logic with the inline algorithm from the C code, which avoids any
+    //  allocations.
+    fn const_word_cmr(v: &Value) -> Cmr {
+        assert_eq!(v.len().count_ones(), 1);
+        let w = 1 + v.len().trailing_zeros() as usize;
+        // 1. compute scribe pass-one IMR
+        let scribe = crate::core::CommitNode::<crate::jet::Core>::scribe(
+            &mut crate::core::Context::new(),
+            v,
+        );
+        let imr_iv = Cmr::tag_iv(b"Simplicity-Draft\x1fIdentity");
+        let imr_pass1 = imr_iv.update_1(scribe.cmr());
+        // 2. Add TMRs to get the pass-two IMR
+        let types = crate::core::types::Type::powers_of_two_vec(15);
+        let imr_pass2 = imr_pass1.update(
+            types[0].tmr.0.into_inner().into(),
+            types[w].tmr.0.into_inner().into(),
+        );
+        // 3. Convert to a jet CMR
+        Cmr::tag_iv(b"Simplicity-Draft\x1fJet").update_1(imr_pass2)
```

Comment:
That is a _big_ fixme. Do you have a link to this inline algorithm? We should open an issue for FFI bindings.

Reply:
Sure: https://github.com/BlockstreamResearch/simplicity/blob/master/C/dag.c#L80-L136

It's not too bad, I think we could basically translate the C code directly into Rust and it'd work. Though we'd still have to figure out how to cache the TMRs.

Reply:
Part of why I did it this way was to double-check that Russell's inline algorithm was actually equivalent to the dumb "just take the IMR of a scribe" version. Committing this will mean we have a copy of that code which is easy to check against.